# 递归函数

## 递归 (recursion)

函数中有个理解门槛比较高的概念：**递归函数**（Recursive）——那些**在自身内部调用自身的函数**。说起来比较拗口。

先看一个例子，我们想要有个能够计算 `n` 的阶乘 `n!` 的函数 `f()`，规则如下：

> * n! = n * (n-1) * (n-2)...* 1
> * 即，n! = n * (n-1)!
> * 且 n >= 1
> 
> **注意**：以上是数学表达式，不是程序，所以，= 在这一小段是“等于”的意思，不是程序语言中的赋值符号。

于是，计算 `f(n)` 的 python 程序如下：

In [1]:
def f(n):
    if n == 1:
        return 1
    else:
        return n * f(n-1)
    
print(f(5))

120


## 1.2 递归函数的执行过程

以 factorial(5) 为例，让我们看看函数的执行流程：

* 5>1, 所以在计算 `5*f(4)` 的时候要再次调用自己，即 f(4)， 所以必须等待 f(4) 的返回值；
* 4>1, 所以在计算 `4*f(3)` 的时候要再次调用自己，即 f(3)， 所以必须等待 f(3) 的返回值；
* 3>1, 所以在计算 `3*f(2)` 的时候要再次调用自己，即 f(2)， 所以必须等待 f(2) 的返回值；
* 2>1, 所以在计算 `2*f(1)` 的时候要再次调用自己，即 f(1)， 所以必须等待 f(1) 的返回值；
* 1==1，所以这个时候就不会调用 `f()` 了，意识递归结束，开始返回，这一次返回值是 1；
* 下一步是返回 `2*f(1)`，即`2*1`；
* 下一步是返回 `3*f(2)`，即`3*2`；
* 下一步是返回 `4*f(3)`，即`4*6`；
* 下一步是返回 `5*f(4)`，即`5*25` —— 至此，外部调用 f(5) 的最终返回值是 120......

加上一些输出语句之后，能更清楚地看到大概的执行流程：

In [3]:
def f(n):
    print('\tn =', n)
    if n == 1:
        print('Returning...')
        print('\tn =', n, 'return:', 1)
        return 1
    else:
        r = n * f(n-1)
        print('\tn =', n, 'return:', r)
        return r
    
print('Call f(5)...')
print('Get out of f(n), and f(5) =', f(5))

Call f(5)...
	n = 5
	n = 4
	n = 3
	n = 2
	n = 1
Returning...
	n = 1 return: 1
	n = 2 return: 2
	n = 3 return: 6
	n = 4 return: 24
	n = 5 return: 120
Get out of f(n), and f(5) = 120


上面这个函数反映出了递归的细节。到底为什么没有先输出第二个 `print` 语句的前半部分，而是先输出函数调用的部分。。。还没搞懂。

## 递归的终点

递归函数在内部必须有一个能够让自己停止调用自己的方式，否则就永远循环下去了。。。

从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，说。。。

写成 python 程序大概是这样：

In [6]:
def a_monk_telling_story():
    print('从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… ')
    return a_monk_telling_story()

a_monk_telling_story()

从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个

从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个老和尚和小和尚，老和尚给小和尚讲故事，他说…… 
从前有座山，山上有座庙，庙里有个

RecursionError: maximum recursion depth exceeded while calling a Python object

这是个**无限循环**的递归，因为这个函数里没有设置中止自我调用的条件。无限循环有个不好听的名字，叫做**死循环**。

盗梦空间 inception，层层嵌套的梦有点类似递归，它设置了中止自我调节的条件：

在电影里，醒过来的条件有两个：
* 在梦里死掉；
* 在梦里被 kicked 到。。。

如果这两个条件一直得不到满足，那就进入了 limbo 状态——就跟死循环一样，出不来了。。。

In [5]:
import random

def in_dream(day=0, dead=False, kicked=False):
    dead = not random.randrange(0,10) # 1/10 probability to be dead
    kicked = not random.randrange(0,10) # 1/10 probability to be kicked
    day += 1
    print('dead:', dead, 'kicked:', kicked)
    
    if dead:
        print((f"I slept {day} days, and was dead to wake up..."))
        return day
    elif kicked:
        print(f"I slept {day} days, and was kicked to wake up...")
        return day
    
    return in_dream(day)
    
print('The in_dream() function returns:', in_dream())

dead: False kicked: False
dead: False kicked: False
dead: True kicked: False
I slept 3 days, and was dead to wake up...
The in_dream() function returns: 3


递归函数的共同特性：

1. 在 `return` 语句中返回的是**对自身的调用**，即，含有自身的表达式。
2. 为了避免死循环，一定要**至少有一个条件下**返回的不再是对自身的调用。

## 变量的作用域

全局变量 globa，和局部变量 local

* 在函数内部被赋值而后使用的，都是局部变量，它的作用域是函数内部，无法被函数外的代码调用；

* 在所有函数之后被赋值而后开始使用的，是全局边阿玲，它们的作用域是全局。在函数内外都可以被调用。

程序员们通常会严格地遵守一条原则：

> 在函数内部绝对不调用全局变量。即使是必须改变全局变量，也只能通过函数的返回值在函数外改变全局变量。

类比到在日常的工作生活中的调用：

> 做事的原则：自己的事自己做；
>
>别人的事，最多通过自己的产出让他们自己去搞。。。。这里没懂。

当一个变量本身被当做参数传递给一个函数的时候，这个变量本身并不会被函数所改变。

In [8]:
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n-1)
    
a = 5
b = factorial(a)   # a 并不会因此改变；
print(a, b)
a = factorial(a)   # 这是你主动为 a 再一次赋值……
print(a, b)

5 120
120 120


递归函数每一次被调用都会形成一个作用域，嵌套的作用域。`n` 作为参数传递进去，但是 `n` 本身并没有被改变。

In [9]:
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n-1)
    
n = 5              # 这一次，这个变量名称是 n
m = factorial(n)   # n 并不会因此改变；
print(n, m)

5 120


为了避免全局变量和局部变量的混淆。最好的策略是不使用重复的变量名。

In [11]:
# 观察一下名称相同的一个全局变量和局部变量的不同内存地址
def f(n):
    return id(n)
    
n = 5
print(id(n))    # 全局变量 n 的内存地址
print(id(f(n))) # 局部变量 n 的内存地址。
print(id(f(f(n))))

140711960875968
2614652285840
2614652285968


## 递归函数三原则

1. 根据定义，递归函数必须在内部调用自己；
2. 必须设定一个退出条件；
3. 递归过程中必须能够逐步达到退出条件；

作为盗梦空间的那个假的递归程序。它满足前两条。但是不是严格满足第三条：
它是**不管怎样都能达到退出条件**，而不是**逐步达到**退出条件。

不完全理解。感觉像是，由于随机过程，可能根本不执行调用自身就完全结束了。

[普林斯顿大学的网站有许多例子](https://introcs.cs.princeton.edu/java/23recursion/)